chore: Add 'grpc.client.deadline' tag to track gRPC client's deadline value

nvh0412 opened the pull request:
<!--
Check out the
https://github.com/DataDog/dd-trace-rb/blob/master/docs/DevelopmentGuide.md
for guidance on how to set up your development environment,
run the test suite, write new integrations, and more.
-->

**What does this PR do?**
resolve https://github.com/DataDog/dd-trace-rb/issues/2199

**Motivation**
I'm using Datadog tracing to methodically trace data from the whole of our internal system (Web server, gRPC, ...). As I use multi-threaded servers to run our gRPC server, hence I'm facing many DeadlineExceed exceptions from the gRPC client ruby because the thread took so long to finish its job, and the timeout setting isn't enough.

Despite talking about the solution, I want to know the value of client timeout when it sends the request to the gRPC server. In this way, I would know how many redundant thread executions are while its client call is terminated because of timeout.

So I'm adding `grpc.client.deadline` tag to track gRPC client's deadline value

TonyCTHsu commented:
Hi @nvh0412 , Thanks for the contribution! 

Could you provide a positive example from spec which verify the `deadline` has been set correctly with a value (`ActiveCall#deadline` would return a `Time` object).

Also please rebase the branch to the latest master to avoid failing spec in old Rubies. 😄

nvh0412 commented:
Thank you @TonyCTHsu , addressed.

nvh0412 commented:
Fair enough, the Time object's serialised value might also affect the Datadog query in our Dashboard. Milliseconds would be better, thanks for suggesting me @TonyCTHsu, addressing.

nvh0412 commented:
Please help me review this PR again, thank you.

marcotc commented on lib/datadog/tracing/contrib/grpc/datadog_interceptor/client.rb:
```diff
@@ -44,6 +44,9 @@ def annotate!(trace, span, metadata, call)
               span.set_tag(Tracing::Metadata::Ext::TAG_PEER_HOSTNAME, host) if host
 
+              deadline = find_deadline(call)
+              span.set_tag(Ext::TAG_CLIENT_DEADLINE, (deadline.to_time.to_f * 1000).to_i) if deadline
```
👋 @nvh0412, could you move the time conversion logic of deadline (from object to number) to the `find_deadline` method?

I think it looks cleaner if the caller here only has to set the deadline, like so:
```suggestion
              span.set_tag(Ext::TAG_CLIENT_DEADLINE, deadline) if deadline
```

marcotc commented on lib/datadog/tracing/contrib/grpc/datadog_interceptor/client.rb:
```diff
@@ -44,6 +44,9 @@ def annotate!(trace, span, metadata, call)
               span.set_tag(Tracing::Metadata::Ext::TAG_PEER_HOSTNAME, host) if host
 
+              deadline = find_deadline(call)
+              span.set_tag(Ext::TAG_CLIENT_DEADLINE, (deadline.to_time.to_f * 1000).to_i) if deadline
```
I'm not sure `to_time` is needed here. It seems like the object is always a `Time` object.

Also, `Time#to_time` is not part of the Ruby standard library:
```sh
$ ruby -e 'Time.at(1).to_time'
-e:1:in `<main>': undefined method `to_time' for 1969-12-31 16:00:01 -0800:Time (NoMethodError)
```
It works in our tests because it seems like some gem is monkey patching `Time`, but we can't guarantee it.

nvh0412 replied:
@marcotc it's actually a `GRPC::Core::TimeSpec` object, not a `Time` object

nvh0412 replied:
![Screenshot from 2022-08-04 07-32-35](https://user-images.githubusercontent.com/1973623/182738391-2bc6443e-cf1d-4f81-baa2-cb72d4168c49.png)

nvh0412 replied:
But, you're right, there is some cases, deadline is a `Time` object, and ActiveSupport is monkey patching `Time`
![Screenshot from 2022-08-04 07-47-20](https://user-images.githubusercontent.com/1973623/182739462-0656fb75-ee43-406a-bceb-1d53f3311a88.png)

let me address it

nvh0412 replied:
Turns out that gRPC gem mentioned that issue in their spec https://github.com/grpc/grpc/blob/3d2e5038c310df3b4d215cb7872a26045e979547/src/ruby/spec/generic/client_stub_spec.rb#L79-L83

So the value of the deadline can be Time or TimeSpec either. That's why I handle 2 cases in this commithttps://github.com/DataDog/dd-trace-rb/pull/2200/commits/f7499de35d6d3dd9fad652cadd385da7ab331fea

marcotc commented on spec/datadog/tracing/contrib/grpc/datadog_interceptor/client_spec.rb:
```diff
@@ -31,9 +31,10 @@
   end
 
   context 'using client-specific configurations' do
+    let(:deadline) { Time.now }
```
I recommend keeping a predictable time for the test suite, something like:
```suggestion
    let(:deadline) { Time.utc(2022, 1, 2, 3, 4, 5, 678901) }
```
For this example, `deadline#to_f` will return `1641092645.678901`, which has the maximum precision supported.

This allows us to test that we received exactly `1641092645678` here:
```ruby
       expect(span.get_tag(Datadog::Tracing::Contrib::GRPC::Ext::TAG_CLIENT_DEADLINE)).to eq(1641092645678)
```

nvh0412 commented:
Hi @TonyCTHsu @marcotc I addressed your feedbacks, also squashed all commit to one, that's why I have to do a push force. Please help me review this PR again, thank you

TonyCTHsu commented:
👋 @nvh0412 
According [grpc](https://grpc.io/blog/deadlines/). There is no true answer for a good timeout. But it seems to be legit that millisecond resolution would be helpful.

> There is no single answer to “What is a good deadline/timeout value?”. Your service might be as simple as the [Greeter](https://github.com/grpc/grpc/blob/master/examples/protos/helloworld.proto) in our quick start guides, in which case 100 ms would be fine. Your service might be as complex as a globally-distributed and strongly consistent database. The deadline for a client query will be different from how long they should wait for you to drop their table.

I took a deeper dive into `grpc` , and yes. The [deadline value](https://github.com/grpc/grpc/blob/master/src/ruby/lib/grpc/core/time_consts.rb#L36) could be a `Time` or `TimeSpec` object. When user provides a positive number, it would ends up to become a `Time` object. It is also possible providing `TimeSpec` and `nil` which return `TimeSpec`. 

I realized that there are three `TimeSpec` object, referenced by `TimeConsts::ZERO` , `TimeConsts::INFINITE_FUTURE` and `TimeConsts::INFINITE_PAST`, when I tried to write test examples for `TimeSpec`. All three of them are referring to value back in 1960s. I wonder whether those `TimeSpec` values are important/useful to your case?

For `Time` serialization, 
```
irb(main):001:0> require 'time'
=> true
irb(main):002:0> Time.now.utc.iso8601(3)
=> "2022-08-04T17:24:09.271Z"
```
This this would provide a human readable and interchangable format, which string comparison should also work for dashboard query.

if `TimeSpec` is not important for visibility. Perhaps we don't need to tag it. But if it is, consider placing a tag `grpc.client.deadline.type`with [TIMESTAMP, INFINITE_FUTURE,  INFINITE_PAST, ZERO] and `grpc.client.deadline.timestamp` with serialized time. This would avoid confusion of ill-formed data of those.

nvh0412 commented:
I don't think `TimeConsts::ZERO`, `TimeConsts:: INFINITE_FUTURE`, `TimeConsts::INFINITE_PAST` are valuable for us. Also, I just figured out that if we tag this `deadline` on the client interceptor, its value is always `Time` ruby, `TimeSpec` only appear on Server interceptor.

So I do believe we can come up with that we don't need to handle `TimeSpec` type in this PR because we only change the client interceptor. But I will address the code that ensures we won't tag the deadline if it's `TimeSpec`.

As for the serialized value of `Time` object, it's a fifty-fifty split `milisecond` number or a string value (iso8601 format). Because the millisecond value would help others who want to calculate/compare with other values easier than a string value, I believe. (i.e duration value), so let's keep it as millisecond for now. Wdyt? @TonyCTHsu

TonyCTHsu commented:
If `TimeSpec` is not interested, let's just skip it. 

I believe I could approve the change after iso8601 format serialization 😄 .

nvh0412 commented:
Addressed @TonyCTHsu , there are also some flaky tests on the master branch https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb?branch=master.

TonyCTHsu reviewed (approved):
@nvh0412  Thanks for the great work! 😘
